### PR TITLE
[MM-41710] Implement pluggable Call button in channel header

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

PR reworks the existing pluggable call button implementation by @larkox and moves it back to the channel header. I purposely tried to keep the logic as much as possible on the plugin's side as the component should allow for display changes based on the state of a call.

Plugin wise this is technically a breaking change but was agreed that the impact should be minimal. Please let me know if there are concerns.

/cc @matthewbirtch @aaronrothschild for visibility.

#### Sample registration

```tsx
const callButton = (
  <button className="style--none call-button">
    <i style={{ color: "#2D8CFF" }} className={`CompassIcon icon-brand-zoom`} />
    <span className="call-button-label">{"Start Zoom meeting"}</span>
  </button>
);
const dropdownButton = (
  <button className={"style--none call-button-dropdown"}>
    <i style={{ color: "#2D8CFF" }} className={`CompassIcon icon-brand-zoom`} />
    <div>
      <span>{"Start Zoom meeting"}</span>
      <span className="call-button-dropdown-sublabel">{"In this channel"}</span>
    </div>
  </button>
);
registry.registerCallButtonAction(callButton, dropdownButton, (channel: Channel) => {
  // do something
});
```

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-41710

#### Design Link

https://www.figma.com/file/U6MKDiNZ6kgW6eV4loXP9K/Audio-rooms?node-id=1208%3A321692

#### Release Note

```release-note
Updated plugin registry's registerCallButtonAction method to allow for displaying custom calls buttons in the channel header.
```

#### Screenshots

**Single registration (Calls plugin)**

- Not in call
image
- In call
image
- With plugins
image

**Multiple registrations**

- Not in call
image
- In call
image
- With plugins
image

### CODE REVIEW

**Review by itao — APPROVED** [review]
Assuming green light from others on grammar and styling, this looks good to me.

**Comment by matthewbirtch** [comment]
Love this! Thanks for tackling it @streamer45. I'll let @abhijit-singh review this, but this is a great addition to the product.

File: plugins/registry.js
Comment (larkox) [inline plugins/registry.js:180]: @hanzei Do we want to pass things here as React elements (both for the button and the dropdown)? This is good since gives more liberty to the plugin developer, but also may cause some problems because of that freedom. We already had some problems with the alignment of the dropdown in the channel header because we were using components styled one way or another.

**Review by larkox — APPROVED** [review]
LGTM, just one concern I will defer to the Integrations team.

**Comment by justinegeffen** [comment]
> Functionally it looks great! @justinegeffen may be able to help us out with grammar ie `Start a call` vs `Start call` and `Join call` vs `Join the call`. It needs to be considered with the subtext `in this channel`, which is also totally open to change - I look to hers and @abhijit-singh's input.

In the Figma mock-up we refer to "Start a call", "Join call", "End call". If we work on the understanding that there can be multiple calls happening on the server simultaneously but only one per channel, then I think that works.

**Comment by mickmister** [comment]
Note that we have an existing issue where if the user clicks the x on the RHS to close the RHS, the mouse is then positioned on the rightmost channel header button (which is now the call button), where the user may accidentally click again if the webapp is lagging a bit. This causes a call to be started by accident.

There is a hacky fix for this that disables the channel header buttons temporarily (500 ms) after the RHS has finished closing:

https://github.com/mattermost/mattermost-webapp/blob/9b93dab6651a5ecf5cb3e8f983142a37c58112ff/plugins/channel_header_plug/channel_header_plug.tsx#L135-L141

@matthewbirtch We have discussed this issue a few times, with the desire for a proper fix without the need for the timeout. Maybe we should consider adding the fix in this PR, possibly repositioning the call button to the left a bit? If we don't have a proper fix here, then we should add the hacky fix here as well.

A customer recently mentioned that this issue is still occurring for them on their server, even with the timeout. They also reported that they have a particularly laggy experience on their server, and we concluded that maybe the timeout needs to be increased. @matthewbirtch You suggested in their thread that we could bump the timeout to 1000ms-1500ms. I suggest we apply that here to avoid a similar issue with this component.

If we do apply the fix here, @streamer45 how would you feel about doing a drive-by fix in the existing channel header components to increase the timeout there as well? I'll create a separate PR for that if you'd like to keep the scope minimal here.

File: components/call_button/call_button.tsx
Comment (mickmister) [inline components/call_button/call_button.tsx:91]: Maybe a good use case for the `classnames` library?

**Review by mickmister — COMMENTED** [review]
Code LGTM :+1:

Though I'm holding off on approving until the discussion of using custom React components is resolved

File: components/call_button/call_button.tsx
Comment (streamer45) [reply components/call_button/call_button.tsx:91]: Clearly I didn't know about its existance, thanks :raised_hands:

**Comment by streamer45** [comment]
> If we do apply the fix here, @streamer45 how would you feel about doing a drive-by fix in the existing channel header components to increase the timeout there as well? I'll create a separate PR for that if you'd like to keep the scope minimal here.

No objections from my side, happy to include any additional fix, once agreed.

**Review by abhijit-singh — APPROVED** [review]
Looks great, thanks @streamer45!

**Comment by matthewbirtch** [comment]
> @matthewbirtch We have discussed this issue a few times, with the desire for a proper fix without the need for the timeout. Maybe we should consider adding the fix in this PR, possibly repositioning the call button to the left a bit? If we don't have a proper fix here, then we should add the hacky fix here as well.
> 
> A customer recently mentioned that this issue is still occurring for them on their server, even with the timeout. They also reported that they have a particularly laggy experience on their server, and we concluded that maybe the timeout needs to be increased. @matthewbirtch You suggested in their thread that we could bump the timeout to 1000ms-1500ms. I suggest we apply that here to avoid a similar issue with this component.
> 
> If we do apply the fix here, @streamer45 how would you feel about doing a drive-by fix in the existing channel header components to increase the timeout there as well? I'll create a separate PR for that if you'd like to keep the scope minimal here.

Yes, let's apply the timeout fix for now. Moving the button isn't viable right now as that will cause misalignment and will look pretty strange.

File: plugins/call_button/call_button.tsx
Comment (mickmister) [inline plugins/call_button/call_button.tsx:51]: I have to say that I'm not well-versed enough in the usage of `useRef` to deem this as correct, though it seems right to me. @michelengelen Would you be able to lend a hand or recommend a developer to review this piece?

**Review by mickmister — APPROVED** [review]
LGTM! I've asked for some help on reviewing the usage of `useRef`

File: plugins/registry.js
Comment (mickmister) [reply plugins/registry.js:180]: I like how it is implemented with the dedicated css classes implemented by the webapp. :+1:

We certainly could have gone that route with the App Bar as well, though it's not set in stone what will be done with the components in the far future so we wanted to be as opinionated as possible on the core webapp side.

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [reply plugins/call_button/call_button.tsx:51]: I am not 100% sure why this is written so complicated. If I understand correctly the purpose of this is to know if the sidebar has been opened before, right? It just seems overly complicated to me to use this IIFE and a `useEffect` call inside to achieve that.

I would suggest to try something like this:

```
    const prevSidebarOpen = useRef(props.sidebarOpen);

    useEffect(() => {
        if (prevSidebarOpen.current && !props.sidebarOpen) {
            setClickEnabled(false);
            setTimeout(() => {
                setClickEnabled(true);
            }, Constants.CHANNEL_HEADER_BUTTON_DISABLE_TIMEOUT);
        }
        prevSidebarOpen.current = props.sidebarOpen;
    }, [props.sidebarOpen]);
```

@mickmister using refs is a good way to store values over renders, without using state. refs do not trigger a re-render when changed, so you can freely manipulate them.

Comment (streamer45) [reply plugins/call_button/call_button.tsx:51]: > I am not 100% sure why this is written so complicated. 

@michelengelen 

That's almost certainly due to my attempt at doing something I didn't know anything about :p
My goal was to have a similar behaviour as to `componentDidUpdate()` where I get the `prevProps` which I want to compare. I googled a bit a around and got the impression I needed to write a custom hook. 

More than happy to simplify of course, thanks for the input there, will give it a try asap.

Comment (michelengelen) [reply plugins/call_button/call_button.tsx]: Feel free to reach out if you need help with the UI 👍🏼

Comment (streamer45) [reply plugins/call_button/call_button.tsx]: @michelengelen Thanks, your suggestion works perfectly fine and it makes sense given we are using it in a single place and we already have a `useEffect` call in place.

File: plugins/call_button/call_button.scss
Comment (michelengelen) [inline plugins/call_button/call_button.scss:140]: @streamer45 Sorry for the massive suggestion, but having them all separated would be even less comprehensible. If you have questions about these proposed changes feel free to reach out.

```suggestion
button.call-button {
    display: flex;
    height: 32px;
    place-items: center;
    place-content: center;
    padding: 10px 12px;
    border-radius: 4px;
    color: rgb(var(--button-bg-rgb));
    fill: currentColor;

    &.dropdown {
        color: rgba(var(--center-channel-color-rgb), 0.56);
        fill: currentColor;
    }
    
    &.active {
        background: rgba(var(--button-bg-rgb), 0.08);
        color: rgb(var(--button-bg-rgb));
        fill: currentColor;
    }

    &:hover {
        background: rgba(var(--button-bg-rgb), 0.08);
        color: rgb(var(--button-bg-rgb));
        fill: currentColor;
    }

    &:active {
        background: rgba(var(--button-bg-rgb), 0.16);
        color: rgb(var(--button-bg-rgb));
        fill: currentColor;
    }
    
    &.disabled {
        color: rgba(var(--center-channel-color-rgb), 0.32);
        fill: currentColor;
    
        &:hover {
            background: inherit;
            color: inherit;
            fill: currentColor;
        }
    
        &:active {
            background: revert;
            color: inherit;
            fill: currentColor;
        }
    }

    i,
    span {
        font-size: 14px;
        line-height: 14px;
    }

    i::before,
    span::before {
        margin: 0;
    }

    svg {
        width: 14px;
        height: 14px;
    }

    .call-button-label {
        padding: 1px 0;
        margin: 0 6px;
        font-size: 12px;
        font-weight: 600;
        line-height: 9px;
    }
}

.Menu .MenuItem > button.call-button-dropdown {
    display: flex;
    padding: 6px 20px 6px 16px;
    margin: 0;
    font-size: 14px;
    line-height: 20px;

    i,
    svg {
        align-self: flex-start;
        margin-right: 12px;
        margin-left: 4px;
    }

    svg {
        width: 16px;
        height: 16px;
        fill: rgb(var(--button-bg-rgb));
    }

    i {
        color: rgb(var(--button-bg-rgb));
        font-size: 16px;

        &::before {
            margin: 0;
        }
    }

    .call-button-dropdown-sublabel {
        margin-top: 4px;
        color: rgba(var(--center-channel-color-rgb), 0.56);
        font-size: 12px;
        line-height: 16px;
    }

    div {
        display: flex;
        flex-direction: column;
    }
    
    &.disabled {
        color: rgba(var(--center-channel-color-rgb), 0.32);
        fill: rgba(var(--center-channel-color-rgb), 0.32);
        
        svg {
            fill: rgba(var(--center-channel-color-rgb), 0.32);
        }
        
        i {
            color: rgba(var(--center-channel-color-rgb), 0.32);
        
            &::before {
                margin: 0;
            }
        }
        
        .call-button-dropdown-sublabel {
            color: rgba(var(--center-channel-color-rgb), 0.32);
            font-size: 12px;
        }
    }
}

```

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [inline plugins/call_button/call_button.tsx:5]: why not use `useIntl`?

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [inline plugins/call_button/call_button.tsx:49]: I am not 100% sure, but you can just pass in `int`s as values, since the standard is to use `px`

```suggestion
            marginTop: 16,
            height: 32,
```

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [inline plugins/call_button/call_button.tsx:56]:
```suggestion
        const item = props.pluginCallComponents[0];
        
        const clickHandler = () => item.action?.(props.currentChannel, props.channelMember);

        return (
```

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [inline plugins/call_button/call_button.tsx:64]:
```suggestion
                onClick={clickEnabled ? clickHandler : undefined}
```

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [inline plugins/call_button/call_button.tsx:69]:
```suggestion
                onTouchEnd={clickEnabled ? clickHandler : undefined}
```

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [inline plugins/call_button/call_button.tsx:85]:
```suggestion
                    item.action?.(props.currentChannel, props.channelMember);
```

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [inline plugins/call_button/call_button.tsx:93]:
```suggestion
    });

    return (
```

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [inline plugins/call_button/call_button.tsx:58]: this would probably more efficient to handle in the scss file, or is there something (dependency, use-case, etc.) I am missing here?

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [inline plugins/call_button/call_button.tsx:102]:
```suggestion
            <MenuWrapper onToggle={(toggle: boolean) => setActive(toggle)}>
```

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [inline plugins/call_button/call_button.tsx:106]: `button`s do by default have `type='button'`

```suggestion
                <button className={classNames('style--none call-button dropdown', {active})}>
```

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [inline plugins/call_button/call_button.tsx:115]: it might be easier to use the IconComponents form the compass-icons package here.

**Review by michelengelen — CHANGES_REQUESTED** [review]
sry for all the comments 🙇🏼

File: plugins/call_button/call_button.tsx
Comment (streamer45) [reply plugins/call_button/call_button.tsx:5]: No idea, the code was (for the most part) already there, I just moved it around and changed the necessary parts. How do these differ?

File: plugins/call_button/call_button.tsx
Comment (streamer45) [reply plugins/call_button/call_button.tsx:58]: My thinking there was that it's a very specific style which doesn't need to be exported elsewhere outside the component.

File: plugins/call_button/call_button.tsx
Comment (streamer45) [reply plugins/call_button/call_button.tsx:115]: Would we be able to retain the aria label as well? I guess we could always wrap them but not sure.

File: plugins/call_button/call_button.scss
Comment (streamer45) [reply plugins/call_button/call_button.scss:140]: Somehow this is not working as expected:

```css
    &.disabled {
        color: rgba(var(--center-channel-color-rgb), 0.32);
        fill: currentColor;
    
        &:hover {
            background: inherit;
            color: inherit;
            fill: currentColor;
        }
    
        &:active {
            background: revert;
            color: inherit;
            fill: currentColor;
        }
    }
```

When hovering the disabled button, the color as defined above (`rgba(var(--center-channel-color-rgb), 0.32)`) is not retained.

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [reply plugins/call_button/call_button.tsx:5]: well, it is cleaner as it just adds a hook and does not wrap the component in a HOC

File: plugins/call_button/call_button.scss
Comment (michelengelen) [reply plugins/call_button/call_button.scss:140]: I might need to check out the PR and test it locally then ... let me do that tomorrow (or maybe later today) and get back here

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [reply plugins/call_button/call_button.tsx:49]: just tested and it does work as expected

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [reply plugins/call_button/call_button.tsx:58]: fair enough

File: components/call_button/call_button.tsx
Comment (streamer45) [reply components/call_button/call_button.tsx]: > Where does "Call Icon" appear in the UI (if at all?) - if it is in the UI, the casing should be "Call icon" which I did change in one of the other files. But it's unclear whether this change is needed so clarity would be appreciated. :)

That's an aria label. It's used by screen readers to know what an element is about. The text is not rendered in the UI. I think we can change it anyway for consistency's sake.

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [reply plugins/call_button/call_button.tsx:5]: yeah, with `useIntl` you don't need to have a prop for it, so that in itself is better. You typically use it like this:

```tsx
const {formatMessage} = useIntl();
```

`formatMessage` is the same function that you are already destructuring from the props, so there is no need to change that part.

**Comment by streamer45** [comment]
@justinegeffen Applied your suggestions with the exception of changes that are unrelated to this PR (other icons). I think those can be addressed in a separate piece of work.

File: plugins/call_button/call_button.tsx
Comment (michelengelen) [inline plugins/call_button/call_button.tsx:9]: nit: could you remove this line please?

**Review by michelengelen — APPROVED** [review]
only one nitpicking comment ... LGTM! 🎉

**Review by michelengelen — APPROVED** [review]

**Comment by streamer45** [comment]
@DHaussermann Please let me know if you need help with this. I am guessing we could run e2e-tests as the changes should only affect the webapp if plugins are registering the provided action. I can also provide sample plugins making the registrations if needed.

**Comment by DHaussermann** [comment]
@streamer45 I can fire off the tests to possibly catch some regression but this tooling will not run the tests with the calls pluign enabled. 
There is still value in doing this now. I can explore the option of me or an SDET possibly running the tests locally with Calls running in a channel header. I will follow up with @saturninoabril on this.

**Comment by streamer45** [comment]
> @streamer45 I can fire off the tests to possibly catch some regression but this tooling will not run the tests with the calls pluign enabled. There is still value in doing this now. I can explore the option of me or an SDET possibly running the tests locally with Calls running in a channel header. I will follow up with @saturninoabril on this.

@DHaussermann  I think the main concern here is making sure webapp is not affected by the changes which shouldn't be the case. I think that testing the plugin itself can be deferred at this point unless it's something you guys can easily do. Let me know, happy to help providing a sample build.

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- High level smoke tested core webapp action
  - Channel and Team creation
  - Switching channel
  - Using different channel types
  - Making posts in main and RHS input box
  - Post display in main and RHS thread
  - Explored autocomplete for input box auto-complete command parser
- Exploratory testing around channel header both with and without calls enabled
- Used multiple plugins that make posts with slack attachments
- Quick test of calls plugin for any new issues 
- Tested on Browser and desktop

No functional issues found
No JS errors observed in console that seem relevant
Based on testing - This PR does also seem to resolve https://mattermost.atlassian.net/browse/MM-42187

LGTM!

**Pull request merged**
